File: java/Fibonacci.java
import java.util.Scanner;

public class Fibonacci {
    // ANSI color codes
    private static final String RESET = "\033[0m";
    private static final String GREEN = "\033[32m";
    private static final String YELLOW = "\033[33m";
    private static final String CYAN = "\033[36m";
    private static final String MAGENTA = "\033[35m";
    private static final String RED = "\033[31m";
    private static final String GRAY = "\033[90m";
    
    // Global variables
    private static long t1, t2, nextTerm, sum;
    private static long maxValue;
    private static double ratio;
    
    // Function to clear screen
    private static void clearScreen() {
        try {
            String os = System.getProperty("os.name").toLowerCase();
            if (os.contains("win")) {
                new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            } else {
                new ProcessBuilder("clear").inheritIO().start().waitFor();
            }
        } catch (Exception e) {
            // If clearing fails, just print newlines
            System.out.print("\033[H\033[2J");
            System.out.flush();
        }
    }
    
    // Procedure to display the header
    private static void displayHeader() {
        System.out.println(GREEN + "===============================================" + RESET);
        System.out.println(GREEN + "      Enhanced Fibonacci Sequence Generator    " + RESET);
        System.out.println(GREEN + "===============================================" + RESET);
        System.out.println();
    }
    
    // Procedure to display menu options
    private static void displayMenu() {
        System.out.println(YELLOW + "Generation Options:" + RESET);
        System.out.println("  1. Generate by number of terms");
        System.out.println("  2. Generate up to a maximum value");
        System.out.println("  3. Display with statistics");
        System.out.println("  4. Display in columns");
        System.out.println();
        System.out.print("Enter your choice (1-4): ");
    }
    
    // Procedure to generate and display Fibonacci sequence
    private static void generateFibonacci(int termCount, boolean showStats, boolean columnMode) {
        t1 = 0;
        t2 = 1;
        sum = 0;
        int termsPerLine = 10;
        
        if (columnMode) {
            System.out.println(CYAN + "Fibonacci Sequence (" + termCount + " terms):" + RESET);
            System.out.println();
        } else {
            System.out.println(CYAN + "Fibonacci Sequence:" + RESET);
        }
        
        // Handle the first term
        if (termCount >= 1) {
            if (columnMode)
                System.out.printf("%12d", t1);
            else
                System.out.print(t1 + " ");
            sum += t1;
        }
        
        // Handle the second term
        if (termCount >= 2) {
            if (columnMode)
                System.out.printf("%12d", t2);
            else
                System.out.print(t2 + " ");
            sum += t2;
        }
        
        // Calculate and print remaining terms
        for (int j = 3; j <= termCount; j++) {
            nextTerm = t1 + t2;
            
            if (columnMode) {
                System.out.printf("%12d", nextTerm);
                if (j % termsPerLine == 0)
                    System.out.println();
            } else {
                System.out.print(nextTerm + " ");
            }
            
            sum += nextTerm;
            
            t1 = t2;
            t2 = nextTerm;
        }
        
        System.out.println();
        System.out.println();
        
        // Display statistics if requested
        if (showStats) {
            System.out.println(MAGENTA + "Statistics:" + RESET);
            System.out.println("  Number of terms: " + termCount);
            System.out.println("  Sum of all terms: " + sum);
            System.out.printf("  Average value: %.2f%n", (double)sum / termCount);
            if (termCount >= 2) {
                System.out.println("  Last term: " + t2);
                System.out.println("  Second-to-last term: " + t1);
                if (t1 != 0) {
                    ratio = (double)t2 / t1;
                    System.out.printf("  Ratio (F(n)/F(n-1)): %.10f%n", ratio);
                    System.out.println("  Golden Ratio (φ): 1.6180339887");
                    System.out.printf("  Difference: %.10f%n", Math.abs(ratio - 1.6180339887));
                }
            }
            System.out.println();
        }
    }
    
    // Procedure to generate up to a maximum value
    private static void generateUpToMax(long maxVal) {
        t1 = 0;
        t2 = 1;
        int count = 0;
        
        System.out.println(CYAN + "Fibonacci Sequence up to " + maxVal + ":" + RESET);
        System.out.println();
        
        if (maxVal >= 0) {
            System.out.print(t1 + " ");
            count++;
        }
        
        if (maxVal >= 1) {
            System.out.print(t2 + " ");
            count++;
        }
        
        while ((t1 + t2) <= maxVal) {
            nextTerm = t1 + t2;
            System.out.print(nextTerm + " ");
            count++;
            t1 = t2;
            t2 = nextTerm;
        }
        
        System.out.println();
        System.out.println();
        System.out.println(MAGENTA + "Total terms generated: " + count + RESET);
        System.out.println();
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n, choice;
        
        // Clear the terminal screen
        clearScreen();
        
        displayHeader();
        displayMenu();
        choice = scanner.nextInt();
        System.out.println();
        
        switch (choice) {
            case 1: {
                System.out.print("Enter the number of terms to generate (max 92): ");
                n = scanner.nextInt();
                System.out.println();
                
                if (n <= 0) {
                    System.out.println(RED + "Error: Please enter a positive integer." + RESET);
                } else if (n > 92) {
                    System.out.println(RED + "Error: Maximum 92 terms supported with long." + RESET);
                    System.out.println("Using 92 terms instead.");
                    n = 92;
                    System.out.println();
                    generateFibonacci(n, false, false);
                } else {
                    generateFibonacci(n, false, false);
                }
                break;
            }
            
            case 2: {
                System.out.print("Enter the maximum value: ");
                maxValue = scanner.nextLong();
                System.out.println();
                
                if (maxValue < 0) {
                    System.out.println(RED + "Error: Please enter a non-negative value." + RESET);
                } else {
                    generateUpToMax(maxValue);
                }
                break;
            }
            
            case 3: {
                System.out.print("Enter the number of terms to generate (max 92): ");
                n = scanner.nextInt();
                System.out.println();
                
                if (n <= 0) {
                    System.out.println(RED + "Error: Please enter a positive integer." + RESET);
                } else if (n > 92) {
                    System.out.println(RED + "Error: Maximum 92 terms supported with long." + RESET);
                    System.out.println("Using 92 terms instead.");
                    n = 92;
                    System.out.println();
                    generateFibonacci(n, true, false);
                } else {
                    generateFibonacci(n, true, false);
                }
                break;
            }
            
            case 4: {
                System.out.print("Enter the number of terms to generate (max 92): ");
                n = scanner.nextInt();
                System.out.println();
                
                if (n <= 0) {
                    System.out.println(RED + "Error: Please enter a positive integer." + RESET);
                } else if (n > 92) {
                    System.out.println(RED + "Error: Maximum 92 terms supported with long." + RESET);
                    System.out.println("Using 92 terms instead.");
                    n = 92;
                    System.out.println();
                    generateFibonacci(n, false, true);
                } else {
                    generateFibonacci(n, false, true);
                }
                break;
            }
            
            default: {
                System.out.println(RED + "Invalid choice. Using default mode (option 1)." + RESET);
                System.out.println();
                System.out.print("Enter the number of terms to generate (max 92): ");
                n = scanner.nextInt();
                System.out.println();
                if (n > 0) {
                    if (n > 92) n = 92;
                    generateFibonacci(n, false, false);
                }
                break;
            }
        }
        
        System.out.println();
        System.out.println(GRAY + "Press ENTER to exit..." + RESET);
        scanner.nextLine(); // consume newline
        scanner.nextLine(); // wait for enter
        scanner.close();
    }
}

File: php/fibonacci.php
#!/usr/bin/env php
<?php

/**
 * Enhanced Fibonacci Sequence Generator in PHP
 */

// ANSI color codes
define('RESET', "\033[0m");
define('GREEN', "\033[32m");
define('YELLOW', "\033[33m");
define('CYAN', "\033[36m");
define('MAGENTA', "\033[35m");
define('RED', "\033[31m");
define('GRAY', "\033[90m");

// Global variables
$t1 = 0;
$t2 = 1;
$nextTerm = 0;
$sum = 0;

/**
 * Clear the terminal screen
 */
function clearScreen() {
    echo "\033[H\033[2J";
    flush();
}

/**
 * Display the program header
 */
function displayHeader() {
    echo GREEN . "===============================================" . RESET . PHP_EOL;
    echo GREEN . "      Enhanced Fibonacci Sequence Generator    " . RESET . PHP_EOL;
    echo GREEN . "===============================================" . RESET . PHP_EOL;
    echo PHP_EOL;
}

/**
 * Display menu options
 */
function displayMenu() {
    echo YELLOW . "Generation Options:" . RESET . PHP_EOL;
    echo "  1. Generate by number of terms" . PHP_EOL;
    echo "  2. Generate up to a maximum value" . PHP_EOL;
    echo "  3. Display with statistics" . PHP_EOL;
    echo "  4. Display in columns" . PHP_EOL;
    echo PHP_EOL;
    echo "Enter your choice (1-4): ";
}

/**
 * Generate and display Fibonacci sequence
 *
 * @param int $termCount Number of terms to generate
 * @param bool $showStats Whether to show statistics
 * @param bool $columnMode Whether to display in columns
 */
function generateFibonacci($termCount, $showStats, $columnMode) {
    global $t1, $t2, $nextTerm, $sum;
    
    $t1 = 0;
    $t2 = 1;
    $sum = 0;
    $termsPerLine = 10;
    
    if ($columnMode) {
        echo CYAN . "Fibonacci Sequence ($termCount terms):" . RESET . PHP_EOL;
        echo PHP_EOL;
    } else {
        echo CYAN . "Fibonacci Sequence:" . RESET . PHP_EOL;
    }
    
    // Handle the first term
    if ($termCount >= 1) {
        if ($columnMode) {
            printf("%12d", $t1);
        } else {
            echo "$t1 ";
        }
        $sum += $t1;
    }
    
    // Handle the second term
    if ($termCount >= 2) {
        if ($columnMode) {
            printf("%12d", $t2);
        } else {
            echo "$t2 ";
        }
        $sum += $t2;
    }
    
    // Calculate and print remaining terms
    for ($j = 3; $j <= $termCount; $j++) {
        $nextTerm = $t1 + $t2;
        
        if ($columnMode) {
            printf("%12d", $nextTerm);
            if ($j % $termsPerLine == 0) {
                echo PHP_EOL;
            }
        } else {
            echo "$nextTerm ";
        }
        
        $sum += $nextTerm;
        
        $t1 = $t2;
        $t2 = $nextTerm;
    }
    
    echo PHP_EOL;
    echo PHP_EOL;
    
    // Display statistics if requested
    if ($showStats) {
        echo MAGENTA . "Statistics:" . RESET . PHP_EOL;
        echo "  Number of terms: $termCount" . PHP_EOL;
        echo "  Sum of all terms: $sum" . PHP_EOL;
        $average = $sum / $termCount;
        printf("  Average value: %.2f\n", $average);
        
        if ($termCount >= 2) {
            echo "  Last term: $t2" . PHP_EOL;
            echo "  Second-to-last term: $t1" . PHP_EOL;
            
            if ($t1 != 0) {
                $ratio = $t2 / $t1;
                $goldenRatio = 1.6180339887;
                $difference = abs($ratio - $goldenRatio);
                printf("  Ratio (F(n)/F(n-1)): %.10f\n", $ratio);
                printf("  Golden Ratio (φ): %.10f\n", $goldenRatio);
                printf("  Difference: %.10f\n", $difference);
            }
        }
        echo PHP_EOL;
    }
}

/**
 * Generate Fibonacci sequence up to a maximum value
 *
 * @param int $maxVal Maximum value to generate up to
 */
function generateUpToMax($maxVal) {
    global $t1, $t2, $nextTerm;
    
    $t1 = 0;
    $t2 = 1;
    $count = 0;
    
    echo CYAN . "Fibonacci Sequence up to $maxVal:" . RESET . PHP_EOL;
    echo PHP_EOL;
    
    if ($maxVal >= 0) {
        echo "$t1 ";
        $count++;
    }
    
    if ($maxVal >= 1) {
        echo "$t2 ";
        $count++;
    }
    
    while (($t1 + $t2) <= $maxVal) {
        $nextTerm = $t1 + $t2;
        echo "$nextTerm ";
        $count++;
        $t1 = $t2;
        $t2 = $nextTerm;
    }
    
    echo PHP_EOL;
    echo PHP_EOL;
    echo MAGENTA . "Total terms generated: $count" . RESET . PHP_EOL;
    echo PHP_EOL;
}

/**
 * Read an integer from standard input
 *
 * @return int|null The integer value or null if invalid
 */
function readInt() {
    $input = trim(fgets(STDIN));
    if (is_numeric($input)) {
        return intval($input);
    }
    return null;
}

/**
 * Main program entry point
 */
function main() {
    clearScreen();
    displayHeader();
    displayMenu();
    
    $choice = readInt();
    if ($choice === null) {
        $choice = 0;
    }
    echo PHP_EOL;
    
    switch ($choice) {
        case 1:
            echo "Enter the number of terms to generate (max 1476): ";
            $n = readInt();
            echo PHP_EOL;
            
            if ($n === null || $n <= 0) {
                echo RED . "Error: Please enter a positive integer." . RESET . PHP_EOL;
            } elseif ($n > 1476) {
                echo RED . "Error: Maximum 1476 terms supported." . RESET . PHP_EOL;
                echo "Using 1476 terms instead." . PHP_EOL;
                $n = 1476;
                echo PHP_EOL;
                generateFibonacci($n, false, false);
            } else {
                generateFibonacci($n, false, false);
            }
            break;
            
        case 2:
            echo "Enter the maximum value: ";
            $maxValue = readInt();
            echo PHP_EOL;
            
            if ($maxValue === null || $maxValue < 0) {
                echo RED . "Error: Please enter a non-negative value." . RESET . PHP_EOL;
            } else {
                generateUpToMax($maxValue);
            }
            break;
            
        case 3:
            echo "Enter the number of terms to generate (max 1476): ";
            $n = readInt();
            echo PHP_EOL;
            
            if ($n === null || $n <= 0) {
                echo RED . "Error: Please enter a positive integer." . RESET . PHP_EOL;
            } elseif ($n > 1476) {
                echo RED . "Error: Maximum 1476 terms supported." . RESET . PHP_EOL;
                echo "Using 1476 terms instead." . PHP_EOL;
                $n = 1476;
                echo PHP_EOL;
                generateFibonacci($n, true, false);
            } else {
                generateFibonacci($n, true, false);
            }
            break;
            
        case 4:
            echo "Enter the number of terms to generate (max 1476): ";
            $n = readInt();
            echo PHP_EOL;
            
            if ($n === null || $n <= 0) {
                echo RED . "Error: Please enter a positive integer." . RESET . PHP_EOL;
            } elseif ($n > 1476) {
                echo RED . "Error: Maximum 1476 terms supported." . RESET . PHP_EOL;
                echo "Using 1476 terms instead." . PHP_EOL;
                $n = 1476;
                echo PHP_EOL;
                generateFibonacci($n, false, true);
            } else {
                generateFibonacci($n, false, true);
            }
            break;
            
        default:
            echo RED . "Invalid choice. Using default mode (option 1)." . RESET . PHP_EOL;
            echo PHP_EOL;
            echo "Enter the number of terms to generate (max 1476): ";
            $n = readInt();
            echo PHP_EOL;
            
            if ($n !== null && $n > 0) {
                if ($n > 1476) $n = 1476;
                generateFibonacci($n, false, false);
            }
            break;
    }
    
    echo PHP_EOL;
    echo GRAY . "Press ENTER to exit..." . RESET . PHP_EOL;
    fgets(STDIN);
}

// Run the main program
main();

